File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/identifier/VariableIdentifier.java
package me.itzisonn_.meazy.parser.ast.expression.identifier;

import kotlin.uuid.Uuid;
import lombok.AllArgsConstructor;
import lombok.Getter;
import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.modifier.Modifiers;
import me.itzisonn_.meazy.runtime.environment.*;
import me.itzisonn_.meazy.runtime.VariableValue;
import me.itzisonn_.meazy.parser.ast.expression.MemberExpression;
import me.itzisonn_.meazy.parser.ast.expression.literal.ThisLiteral;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.lang.constant.ClassDesc;

@NullMarked
public class VariableIdentifier extends Identifier {
    public VariableIdentifier(String id) {
        super(id);
    }

    @Override
    public void emit(InstructionsSet instructions, Environment environment, ProgramUnit parent) {
        ResolvedVariable resolvedVariable = resolveVariable(environment, parent);

        if (resolvedVariable.getClassDesc() == null) {
            instructions.getLocal(resolvedVariable.getType(), resolvedVariable.getSlot());
        }
        else if (resolvedVariable.getTarget() == null) {
            instructions.getStaticField(resolvedVariable.getClassDesc(), resolvedVariable.getId(), resolvedVariable.getType());
        }
        else {
            resolvedVariable.getTarget().emit(instructions, environment, this);
            Uuid endLabel = null;

            if (parent instanceof MemberExpression memberExpression) {
                if (!memberExpression.isNullSafe()) {
                    if (resolvedVariable.getTarget().getType(environment, this).isNullable()) {
                        throw new RuntimeException("Unsafe member call of function " + id + " on object of type " + resolvedVariable.getClassDesc().descriptorString());
                    }
                }
                else {
                    var nonnullLabel = instructions.createAndInitLabel();
                    endLabel = instructions.createAndInitLabel();

                    instructions.duplicate();
                    instructions.gotoLabelIfNonNull(nonnullLabel);

                    instructions.pop();
                    instructions.loadNull();
                    instructions.gotoLabel(endLabel);

                    instructions.bindLabel(nonnullLabel);
                }
            }

            instructions.getField(resolvedVariable.getClassDesc(), resolvedVariable.getId(), resolvedVariable.getType());

            if (endLabel != null) {
                instructions.bindLabel(endLabel);
            }
        }
    }

    @Override
    public DataType getType(Environment environment, ProgramUnit parent) {
        ResolvedVariable resolvedVariable = resolveVariable(environment, parent);
        return DataType.of(resolvedVariable.getType(), resolvedVariable.isNullable());
    }

    private ResolvedVariable resolveVariable(Environment environment, ProgramUnit parent) {
        VariableValue variableValue = resolveMeazyVariable(environment, parent);
        if (variableValue == null) throw new RuntimeException("Can't find variable " + id);

        String className = variableValue.getParentEnvironment().getFullClassName();

        Expression target;
        if (parent instanceof MemberExpression memberExpression) {
            target = memberExpression.getReceiver() instanceof ClassIdentifier ? null : memberExpression.getReceiver();
        }
        else if (variableValue.getModifiers().contains(Modifiers.SHARED()) || variableValue.getParentEnvironment() instanceof FileEnvironment) {
            target = null;
        }
        else {
            target = new ThisLiteral();
        }

        return new ResolvedVariable(
                className == null ? null : ClassDesc.of(className),
                className == null ? variableValue.getSlot() : -1,
                variableValue.getId(),
                variableValue.getDataType().getClassDesc(),
                variableValue.getDataType().isNullable(),
                target
        );
    }

    @Nullable
    private VariableValue resolveMeazyVariable(Environment environment, ProgramUnit parent) {
        String id = getId();

        if (parent instanceof MemberExpression memberExpression) {
            DataType dataType = memberExpression.getReceiver().getType(environment, this);
            ClassDesc classDesc = dataType.getClassDesc();

            ClassEnvironment classEnvironment = EnvironmentUtils.getClass(environment, classDesc).orElse(null);
            if (classEnvironment == null) return null;

            return classEnvironment.getVariable(id).orElse(null);
        }

        return EnvironmentUtils.getVariable(environment, id).orElse(null);
    }

    @Getter
    @AllArgsConstructor
    private static class ResolvedVariable {
        @Nullable
        private final ClassDesc classDesc;
        private final int slot;
        private final String id;
        private final ClassDesc type;
        private final boolean isNullable;
        @Nullable
        private final Expression target;
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/identifier/VariableIdentifier.kt
package me.itzisonn_.meazy.parser.ast.expression.identifier

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.DataType
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.parser.ast.expression.MemberExpression
import me.itzisonn_.meazy.parser.ast.expression.literal.ThisLiteral
import me.itzisonn_.meazy.parser.modifier.Modifiers
import me.itzisonn_.meazy.runtime.VariableValue
import me.itzisonn_.meazy.runtime.environment.Environment
import me.itzisonn_.meazy.runtime.environment.FileEnvironment
import me.itzisonn_.meazy.runtime.environment.getClass
import me.itzisonn_.meazy.runtime.environment.getVariable
import java.lang.constant.ClassDesc
import kotlin.uuid.Uuid

class VariableIdentifier(id: String) : Identifier(id) {
    override fun emit(instructions: InstructionsSet, environment: Environment, parent: ProgramUnit) {
        val resolvedVariable = resolveVariable(environment, parent)

        if (resolvedVariable.classDesc == null) {
            instructions.getLocal(resolvedVariable.type, resolvedVariable.slot)
        }
        else if (resolvedVariable.target == null) {
            instructions.getStaticField(resolvedVariable.classDesc, resolvedVariable.id, resolvedVariable.type)
        }
        else {
            resolvedVariable.target.emit(instructions, environment, this)
            var endLabel: Uuid? = null

            if (parent is MemberExpression) {
                if (!parent.isNullSafe) {
                    if (resolvedVariable.target.getType(environment, this).isNullable) {
                        throw RuntimeException("Unsafe member call of function $id on object of type ${resolvedVariable.classDesc.descriptorString()}")
                    }
                }
                else {
                    val nonnullLabel = instructions.createAndInitLabel()
                    endLabel = instructions.createAndInitLabel()

                    instructions.duplicate()
                    instructions.gotoLabelIfNonNull(nonnullLabel)

                    instructions.pop()
                    instructions.loadNull()
                    instructions.gotoLabel(endLabel)

                    instructions.bindLabel(nonnullLabel)
                }
            }

            instructions.getField(resolvedVariable.classDesc, resolvedVariable.id, resolvedVariable.type)

            if (endLabel != null) {
                instructions.bindLabel(endLabel)
            }
        }
    }

    override fun getType(environment: Environment, parent: ProgramUnit): DataType {
        val resolvedVariable = resolveVariable(environment, parent)
        return DataType.of(resolvedVariable.type, resolvedVariable.isNullable)
    }

    private fun resolveVariable(environment: Environment, parent: ProgramUnit): ResolvedVariable {
        val variableValue = resolveMeazyVariable(environment, parent) ?: error("Can't find variable $id")

        val className = variableValue.parentEnvironment.fullClassName
        val target = if (parent is MemberExpression) {
            if (parent.receiver is ClassIdentifier) null else parent.receiver
        }
        else if (Modifiers.shared in variableValue.modifiers || variableValue.parentEnvironment is FileEnvironment) {
            null
        }
        else {
            ThisLiteral()
        }

        return ResolvedVariable(
            if (className == null) null else ClassDesc.of(className),
            if (className == null) variableValue.slot else -1,
            variableValue.id!!,
            variableValue.dataType.classDesc,
            variableValue.dataType.isNullable,
            target
        )
    }

    private fun resolveMeazyVariable(environment: Environment, parent: ProgramUnit): VariableValue? {
        if (parent is MemberExpression) {
            val dataType = parent.receiver.getType(environment, this)
            val classDesc = dataType.classDesc

            val classEnvironment = environment.getClass(classDesc) ?: return null
            return classEnvironment.getVariable(id).orElse(null)
        }

        return environment.getVariable(id)
    }

    private class ResolvedVariable(
        val classDesc: ClassDesc?,
        val slot: Int,
        val id: String,
        val type: ClassDesc,
        val isNullable: Boolean,
        val target: Expression?
    )
}